### PR TITLE
Fix langchain input conversion

### PR DESCRIPTION
**Pull request by serena-ruan**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/serena-ruan/mlflow/pull/13725?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13725/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13725
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
Fix langchain input data conversion, as there's a bug report showing that langchain model's input_schema is not reliable for validating if the model can accept List[BaseMessage] as input:
image
Updated the logic to happen inside APIRequest, so we always try invoking with the List[BaseMessage] input if the input is likely a ChatCompletionRequest. If it fails we fallback to original request.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests
The added test fails on master branch

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

**Review by harupy — DISMISSED** [review]
LGTM

File: mlflow/langchain/api_request_parallel_processor.py [hunk 1]
```diff
@@ -107,23 +108,64 @@ class APIRequest:
...
+                    )
+                result = self._call_model(
+                    request_json, config, callback_handlers, stream=self.stream, **kwargs
+                )
+                # only update after above execution succeeds to make sure error message is correct
+                self.request_json = request_json
+                did_perform_chat_conversion = True
+            # it is hard to catch specific exceptions here because langchain models
+            # don't raise common input validation errors
+            # TODO: what if the model accepts the messages input but just fails?
```
Comment (serena-ruan) [inline mlflow/langchain/api_request_parallel_processor.py:155]: This is the risk though, but I haven't come up with a better solution. In the old ways, I think we just convert the request to list[message] and try to invoke with it directly, which is also not accurate

File: mlflow/langchain/api_request_parallel_processor.py [hunk 2]
```diff
@@ -107,23 +108,64 @@ class APIRequest:
...
+            except Exception as e:
+                _logger.debug(
+                    f"Invoking model with chat messages failed. Error: {e!r}. ", stack_info=True
```
Comment (harupy) [inline mlflow/langchain/api_request_parallel_processor.py:158]:
```suggestion
                    f"Invoking model with chat messages failed. Error: {e!r}. ", exc_info=True
```

File: mlflow/langchain/api_request_parallel_processor.py [hunk 3]
```diff
@@ -107,23 +108,64 @@ class APIRequest:
...
+    def _predict_single_input(
+        self, callback_handlers: Optional[list[BaseCallbackHandler]], **kwargs
+    ):
+        config = kwargs.pop("config", {})
+        config["callbacks"] = config.get("callbacks", []) + (callback_handlers or [])
+        if json_dict_might_be_chat_request(self.request_json):
+            # Try to convert the request to list[BaseMessage] type and call the model
+            # If the conversion fails, call the model with the original request
+            # Note: we cannot rely on the model's input_schema to validate the request
+            # because lots of cases the input_schema is not reliable.
```
Comment (harupy) [inline mlflow/langchain/api_request_parallel_processor.py:138]: Can you add an example here?

File: mlflow/langchain/api_request_parallel_processor.py [hunk 4]
```diff
@@ -107,23 +108,64 @@ class APIRequest:
...
+                request_json = _convert_chat_request_to_messages_or_throw(self.request_json)
+                if self.stream:
+                    # validate if the model accepts list[BaseMessage] as input
+                    # as stream doesn't calculate the result immediately
+                    self._call_model(
+                        request_json, config, callback_handlers, stream=False, **kwargs
+                    )
+                result = self._call_model(
+                    request_json, config, callback_handlers, stream=self.stream, **kwargs
+                )
```
Comment (harupy) [inline mlflow/langchain/api_request_parallel_processor.py:149]: Can this lead to unnecessary token usage?

File: mlflow/langchain/utils/chat.py [hunk 5]
```diff
@@ -213,7 +213,13 @@ def _convert(chunk):
     return map(_convert, chunk_iter)
 
 
-def _convert_chat_request_or_throw(chat_request: dict):
+def _convert_chat_request_to_messages_or_throw(chat_request: dict) -> list[BaseMessage]:
```
Comment (harupy) [inline mlflow/langchain/utils/chat.py:216]: when does this function throw?

Comment (serena-ruan) [reply mlflow/langchain/utils/chat.py:216]: When convert_to_messages fails, e.g. the dictionary doesn't follow {"role": ..., "content": ...} format, or the role is not supported

File: mlflow/langchain/api_request_parallel_processor.py (hunk 4 above)
Comment (serena-ruan) [reply mlflow/langchain/api_request_parallel_processor.py:151]: It depends how token usage is calculated. If the callback only captures a single call then it's not impacted

File: mlflow/langchain/api_request_parallel_processor.py [hunk 6]
```diff
@@ -107,23 +108,66 @@ class APIRequest:
...
+            # because lots of cases the input_schema is not reliable.
+            # e.g. model defined in test test_pyfunc_converts_chat_request_correctly
+            # accepts list[BaseMessage] but the input_schema cannot validate it.
+            try:
+                request_json = _convert_chat_request_to_messages_or_throw(self.request_json)
+                if self.stream:
+                    # validate if the model accepts list[BaseMessage] as input
+                    # as stream doesn't calculate the result immediately
+                    self._call_model(
+                        request_json, config, callback_handlers, stream=False, **kwargs
```
Comment (harupy) [inline mlflow/langchain/api_request_parallel_processor.py:147]: I this significantly slows down the response.

**Pull request closed without merging**